walletnode/services/userdataprocess/task.go

```diff
@@ -0,0 +1,391 @@
+		} else {
+			// Get block hash string
+			blockInfo, err := task.pastelClient.GetBlockVerbose1(ctx, blockNum)
+			if err != nil {
+				log.WithContext(ctx).Debug("failed to get block info with error: %w", err)
+			} else {
+				blockHash = blockInfo.Hash
+			}
+		}
+
+		task.request.PreviousBlockHash = blockHash
+
+		// Get the value of task.request.ArtistPastelIDPassphrase for sign data, then empty it in the request to make sure it not sent to supernodes
+		passphrase := task.request.ArtistPastelIDPassphrase
+		task.request.ArtistPastelIDPassphrase = ""
+		// Marshal task.request to byte array for signing
+		js, err := json.Marshal(task.request)
+		if err != nil {
+			return errors.Errorf("failed to encode request %w", err)
+		}
+
+		// Hash the request
+		hashvalue, err := userdata.Sha3256hash(js)
+		if err != nil {
+			return errors.Errorf("failed to hash request %w", err)
+		}
+
+		// Sign request with Wallet Node's pastelID and passphrase
+		signature, err := task.pastelClient.Sign(ctx, hashvalue, task.request.ArtistPastelID, passphrase, "ed448")
+		if err != nil {
+			return errors.Errorf("failed to sign ticket %w", err)
+		}
+
+		userdata := &userdata.ProcessRequestSigned{
+			Userdata:     task.request,
+			UserdataHash: hex.EncodeToString(hashvalue),
+			Signature:    hex.EncodeToString(signature),
+		}
+
+		// Send userdata to supernodes for storing in MDL's rqlite db.
+		if err := nodes.SendUserdata(ctx, userdata); err != nil {
+			return err
+		}
+		res, err := task.AggregateResult(ctx, nodes)
+		if err != nil {
+			return err
+		}
+		// Post on result channel
+		task.resultChan <- &res
+		log.WithContext(ctx).WithField("userdata_result", res).Debug("Posted userdata result")
+	}
+
+	// close the connections
+	for i := range nodes {
+		if err := nodes[i].Connection.Close(); err != nil {
+			return errors.Errorf("failed to close connection to node %s %w", nodes[i].PastelID(), err)
+		}
+	}
+
+	return nil
+}
+
+// AggregateResult aggregate all results return by all supernode, and consider it valid or not
+func (task *Task) AggregateResult(_ context.Context, nodes node.List) (userdata.ProcessResult, error) {
+	// There is following common scenarios when supernodes response:
+	// 1. Secondary node and primary node both response userdata validation result error
+	// 2. Secondary node response userdata validation result success and primary node provide further processing result
+	// 3. Some node fail to response, or not in the 2 case above, then we need to aggregate result and consider what happen
+
+	// This part is for case 1 or 2 above, and we trust the primary node so we use its response
+	for _, node := range nodes {
+		node := node
+		if node.IsPrimary() {
+			result := node.Result
+			if result == nil {
+				return userdata.ProcessResult{}, errors.Errorf("Primary node have empty result")
+			}
+			return *result, nil
+		}
+	}
+
+	// This part of aggregate the response for case 3 and for future use
```

Comment:
Is this intended for the next phase ? if not please remove the commented code.

Reply:
deleted
thank you